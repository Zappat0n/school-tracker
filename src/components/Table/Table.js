import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import requestTable from './TableQueries';
import { addTable, changeTitle } from '../../slices/userSlice';
import './Table.scss';

const Table = ({ tableName, id, title }) => {
  const request = `${tableName}${id ? `/${id}` : ''}`;
  const table = useSelector((state) => state.user.tables[request]);
  const token = useSelector((state) => state.user.token);
  const history = useHistory();
  const dispatch = useDispatch();

  dispatch(changeTitle(title));

  async function query() {
    if (!token) return;
    const response = await requestTable(request, token);
    if (response && response.data) dispatch(addTable(response));
  }

  if (!table) query();

  const filterKeys = (array) => array.filter((key) => (!['id', 'created_at', 'updated_at'].includes(key) && !key.endsWith('_id')));

  const redirectTo = (rowId) => {
    if (!id) history.push(`/${tableName}${rowId ? `/${rowId}/` : ''}`);
  };

  const formatColumn = (text) => (text.charAt(0).toUpperCase() + text.slice(1).toLowerCase()).replace('_', ' ');

  const addData = () => {
    const response = table.data.map((row) => (
      <tr key={row.id} className="row" onClick={() => redirectTo(row.id)}>
        {filterKeys(Object.keys(row)).map((key, index) => (<td className={`column${index + 1}`} key={key}>{row[key]}</td>))}
      </tr>
    ));
    return response;
  };

  return (
    <div className="container-table">
      {(table && table.data) ? (
        <div className="table">
          <table>
            <thead className="table-head">
              <tr>
                {filterKeys(Object.keys(table.data[0])).map((value, index) => (
                  <th className={`column${index + 1}`} key={value}>{formatColumn(value)}</th>
                ))}
              </tr>
            </thead>
            <tbody className="table-body js-pscroll">
              {addData()}
            </tbody>
          </table>
        </div>
      ) : ''}
    </div>
  );
};

Table.defaultProps = {
  id: undefined,
};

Table.propTypes = {
  tableName: PropTypes.string.isRequired,
  id: PropTypes.string,
  title: PropTypes.string.isRequired,
};

export default Table;
