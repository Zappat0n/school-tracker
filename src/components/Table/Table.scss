@import "../../index.scss";

.container-table {
  @include responsive;

  margin: 0 auto;
  background: #fff;
}

.table-head {
  width: 100%;
  margin-bottom: 2rem;
}

table {
  width: 100%;
  border-spacing: 0;
}

thead,
tbody {
  width: 100%;
}

.table {
  margin: 2rem auto;
  border-radius: 1rem;
  font-family: 'Free Sans', sans-serif;
  background-color: #fff;
  width: 100%;
  overflow: hidden;
  -moz-box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);
  -webkit-box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);
  -o-box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);
  -ms-box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);
  box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);

  th {
    @include big-font;

    color: #fff;
    text-align: center;
    background-color: $primary-blue;
  }

  td {
    @include small-font;

    text-align: center;
  }

  .table-body tr:nth-child(even) {
    background-color: #f8f6ff;
  }
}

.row:hover {
  background: lightgray !important;
}

.column1 {
  padding-left: 0.5rem;
}
