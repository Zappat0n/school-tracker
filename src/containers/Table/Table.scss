@import "../../index.scss";

.container-table {
  width: 80%;
  margin: 0 auto;
  background: #fff;
}

.table-head {
  width: 100%;
  margin-bottom: 2rem;
}

table {
  width: 100%;
  border-spacing: 0;
}

thead,
tbody {
  width: 100%;
}

.table {
  margin: 2rem auto;
  border-radius: 1rem;
  font-family: 'Free Sans', sans-serif;
  background-color: #fff;
  width: 100%;
  overflow: hidden;
  -moz-box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);
  -webkit-box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);
  -o-box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);
  -ms-box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);
  box-shadow: 0 0 40px 0 rgba(0, 0, 0, 0.15);

  th {
    font: normal normal 700 1.5rem/3rem 'Free Sans', sans-serif;
    color: #fff;
    background-color: $primary-blue;
  }

  td {
    font: normal normal 400 1.2rem/2rem 'Free Sans', sans-serif;
    text-align: center;
  }

  .table-body tr:nth-child(even) {
    background-color: #f8f6ff;
  }
}

.row:hover {
  background: lightgray !important;
  cursor: pointer;
}

.column1 {
  padding-left: 2rem;
}
