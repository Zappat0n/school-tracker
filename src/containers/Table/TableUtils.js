const filterKeys = (array) => array.filter((key) => (!['id', 'created_at', 'updated_at'].includes(key) && !key.endsWith('_id')));

const getCommands = (table) => {
  if (table.startsWith('classrooms')) {
    if (table.includes('/')) {
      return [{ name: 'scores', route: '/school-tracker/students/:id' }];
    }
    return [
      { name: 'list', route: '/school-tracker/classrooms/:id' },
      { name: 'scores', route: '/school-tracker/events/:id' },
    ];
  }

  if (table.startsWith('presentation_areas')) {
    if (table.includes('/')) {
      return [
        { name: 'list', route: '/school-tracker/presentation_subareas/:id' },
      ];
    }
    return [
      { name: 'list', route: '/school-tracker/presentation_areas/:id' },
    ];
  }

  if (table.startsWith('presentation_subareas')) {
    if (table.includes('/')) return [];
    return [
      { name: 'list', route: '/school-tracker/presentation_subareas/:id' },
    ];
  }

  if (table.includes('/')) return [];
  return [
    { name: 'list', route: '' },
  ];
};

export { getCommands, filterKeys };
