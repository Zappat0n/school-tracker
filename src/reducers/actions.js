export const SET_USERNAME = 'SET_USERNAME';
export const SET_TITLE = 'SET_TITLE';

export const SAVE_SCORE = 'SAVE_SCORE';
export const SAVE_SCORES = 'SAVE_SCORES';

export const saveUsername = (username) => ({
  type: SET_USERNAME,
  payload: username,
});

export const changeTitle = (title) => ({
  type: SET_TITLE,
  payload: title,
});

/* ------ */

export const saveScore = (score) => ({
  type: SAVE_SCORE,
  payload: score,
});

export const saveScores = (scores) => ({
  type: SAVE_SCORES,
  payload: scores,
});
