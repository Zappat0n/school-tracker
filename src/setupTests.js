// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import { createStore, combineReducers } from 'redux';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import userReducer from './slices/userSlice';
import classroomTableReducer from './slices/classroomTableSlice';

const rootReducer = combineReducers({
  user: userReducer, classroomTable: classroomTableReducer,
});

const initialState = {
  user: {
    tables: {},
    title: '',
  },
  classroomTable: {
    students: [],
    presentations: [],
    scores: {},
  },
};

const store = createStore(rootReducer, initialState);

const renderComponent = (component) => render(
  <Provider store={store}>
    <BrowserRouter>
      {component}
    </BrowserRouter>
  </Provider>,
);

const rendererForComponent = (component) => renderer.create(
  <Provider store={store}>
    <BrowserRouter>
      {component}
    </BrowserRouter>
  </Provider>,
);

export { renderComponent, rendererForComponent, store };
