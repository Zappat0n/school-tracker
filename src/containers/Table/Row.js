import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { filterKeys } from './TableUtils';

const Row = (props) => {
  const { data, commands } = props;

  const drawButtons = () => (
    <td className="buttons">
      {commands.map((command) => (
        <Link key={command.name} to={command.route.replace(':id', data.id)} className="row-link">{command.name}</Link>
      ))}
    </td>
  );

  return (
    <>
      <tr key={data.id} className="row">
        {filterKeys(Object.keys(data)).map((key, index) => (
          <td className={`column${index + 1}`} key={key}>
            {data[key]}
          </td>
        ))}
        {commands.length !== 0 && drawButtons()}
      </tr>
    </>
  );
};

Row.propTypes = {
  data: PropTypes.instanceOf(Object).isRequired,
  commands: PropTypes.instanceOf(Array).isRequired,
};

export default Row;
