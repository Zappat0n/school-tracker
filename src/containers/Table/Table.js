import PropTypes from 'prop-types';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import requestTable from './TableQueries';
import { addTable, changeTitle, setError } from '../../slices/userSlice';
import { filterKeys, getCommands } from './TableUtils';
import Row from './Row';
import './Table.scss';

const Table = ({ tableName, id, title }) => {
  const request = `${tableName}${id ? `/${id}` : ''}`;
  const commands = getCommands(request);
  const table = useSelector((state) => state.user.tables[request]);
  const token = useSelector((state) => state.user.token);
  const dispatch = useDispatch();

  async function query() {
    if (!token) return;
    const response = await requestTable(request, token);
    if (response && response.data) dispatch(addTable(response));
    else dispatch(setError(response.errors));
  }

  if (!table) query();

  const formatColumn = (text) => (text.charAt(0).toUpperCase() + text.slice(1).toLowerCase()).replace('_', ' ').trim();

  useEffect(() => {
    dispatch(changeTitle(title));
  }, []);

  return (
    <div className="container-table">
      {table && table.data && table.data.length > 0 && (
        <div className="table">
          <table>
            <thead className="table-head">
              <tr>
                {filterKeys(Object.keys(table.data[0])).map((value, index) => (
                  <th className={`column${index + 1}`} key={value}>{formatColumn(value)}</th>
                ))}
                {commands.length > 0 && <th key="commands">Commands</th>}
              </tr>
            </thead>
            <tbody className="table-body js-pscroll">
              {table.data.map(
                (row) => (
                  <Row
                    key={row.id}
                    data={row}
                    commands={commands}
                  />
                ),
              )}
            </tbody>
          </table>
        </div>
      )}
    </div>
  );
};

Table.defaultProps = {
  id: undefined,
};

Table.propTypes = {
  tableName: PropTypes.string.isRequired,
  id: PropTypes.string,
  title: PropTypes.string.isRequired,
};

export default Table;
