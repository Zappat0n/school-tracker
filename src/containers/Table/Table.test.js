import React from 'react';
import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import { renderComponent, rendererForComponent, store } from '../../setupTests';
import Table from './Table';
import { addTable } from '../../reducers/actions';

test('Display title', () => {
  store.dispatch(addTable({
    tableName: 'classrooms/1',
    headers: 'headers',
    data: [
      {
        id: 1,
        name: 'John',
        birth_date: '2018-01-01',
      },
      {
        id: 2,
        name: 'Peter',
        birth_date: '2018-01-01',
      },
      {
        id: 3,
        name: 'Mark',
        birth_date: '2018-01-01',
      },
      {
        id: 4,
        name: 'Louis',
        birth_date: '2018-01-01',
      },
    ],
  }));

  renderComponent(
    <>
      <Table tableName="classrooms" id="1" title="Classrooms" />
    </>,
  );

  expect(store.getState().userReducer.tables['classrooms/1'].data.length).toBe(4);
  expect(screen.getByText('Name')).toBeInTheDocument();
  expect(screen.getByText('John')).toBeInTheDocument();
  expect(screen.getByText('Peter')).toBeInTheDocument();
  expect(screen.getByText('Mark')).toBeInTheDocument();
  expect(screen.getByText('Louis')).toBeInTheDocument();
});

test('Table Match SnapShot', () => {
  const tree = rendererForComponent(
    <>
      <Table tableName="classrooms" id="1" title="Classrooms" />
    </>,
  );
  expect(tree).toMatchSnapshot();
});
