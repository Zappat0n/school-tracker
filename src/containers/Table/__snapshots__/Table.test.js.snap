// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Match SnapShot 1`] = `
<div
  className="container-table"
>
  <div
    className="table"
  >
    <table>
      <thead
        className="table-head"
      >
        <tr>
          <th
            className="column1"
          >
            Name
          </th>
          <th
            className="column2"
          >
            Birth date
          </th>
          <th>
            Commands
          </th>
        </tr>
      </thead>
      <tbody
        className="table-body js-pscroll"
      >
        <tr
          className="row"
        >
          <td
            className="column1"
          >
            John
          </td>
          <td
            className="column2"
          >
            2018-01-01
          </td>
          <td
            className="buttons"
          >
            <a
              className="row-link"
              href="/school-tracker/students/1"
              onClick={[Function]}
            >
              scores
            </a>
          </td>
        </tr>
        <tr
          className="row"
        >
          <td
            className="column1"
          >
            Peter
          </td>
          <td
            className="column2"
          >
            2018-01-01
          </td>
          <td
            className="buttons"
          >
            <a
              className="row-link"
              href="/school-tracker/students/2"
              onClick={[Function]}
            >
              scores
            </a>
          </td>
        </tr>
        <tr
          className="row"
        >
          <td
            className="column1"
          >
            Mark
          </td>
          <td
            className="column2"
          >
            2018-01-01
          </td>
          <td
            className="buttons"
          >
            <a
              className="row-link"
              href="/school-tracker/students/3"
              onClick={[Function]}
            >
              scores
            </a>
          </td>
        </tr>
        <tr
          className="row"
        >
          <td
            className="column1"
          >
            Louis
          </td>
          <td
            className="column2"
          >
            2018-01-01
          </td>
          <td
            className="buttons"
          >
            <a
              className="row-link"
              href="/school-tracker/students/4"
              onClick={[Function]}
            >
              scores
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
